design/tracing-design.md

```diff
@@ -0,0 +1,142 @@
+
+## High-Level Design
+The Contour configuration file will be extended with a new optional `tracing` section.
+This configuration block, if present, will enable tracing and will define the trace format and other properties needed to generate and export trace data.
+Initially, the `opencensus` and `zipkin` trace formats will be supported.
+
+The parameters defined in the configuration file will be used to configure each [HTTP connection manager's `tracing` section][5].
+
+Since the tracing backend is pluggable, Contour will not package any particular backend.
+The user is responsible for deploying and operating the tracing backend of their choice, and configuring Contour to make use of it.
+
+## Detailed Design
+The `tracing` section of the Contour config file will look like:
+
+(WIP, this is the rough idea)
+```yaml
+tracing:
+  # Trace format to use. Valid options are opencensus and zipkin.
+  format: [opencensus, zipkin]
+  config:
+    # If tracing.format == opencensus, then this block is expected
+    # to be populated.
+    opencensus:
+      agent-address: otel-collector.default:55678
+    # If tracing.format == zipkin, then this block is expected
+    # to be populated.
+    zipkin:
+      collector-cluster: <Envoy cluster name>
+      collector-endpoint: /api/v2/spans
+      collector-endpoint-version: ...
+```
+
+In the presence of the `tracing` block above, all HTTP connection managers (HCM) will be configured for tracing.
+
+A Zipkin-enabled HCM will include the following as an example:
+```go
+...
+Tracing: &http.HttpConnectionManager_Tracing{
+    Provider: &envoy_config_trace_v3.Tracing_Http{
+        Name: "envoy.tracers.zipkin",
+        ConfigType: &envoy_config_trace_v3.Tracing_Http_TypedConfig{
+            TypedConfig: protobuf.MustMarshalAny(&envoy_config_trace_v3.ZipkinConfig{
+                CollectorCluster:         "default/zipkin/9411/da39a3ee5e",
+                CollectorEndpoint:        "/api/v2/spans",
+                CollectorEndpointVersion: envoy_config_trace_v3.ZipkinConfig_HTTP_JSON,
+            }),
+        },
+    },
+},
+...
+```
+
+An OpenCensus-enabled HCM will include the following as an example:
+```go
+...
+Tracing: &http.HttpConnectionManager_Tracing{
+    Provider: &envoy_config_trace_v3.Tracing_Http{
+    Name: "envoy.tracers.opencensus",
+    ConfigType: &envoy_config_trace_v3.Tracing_Http_TypedConfig{
+        TypedConfig: protobuf.MustMarshalAny(&envoy_config_trace_v3.OpenCensusConfig{
+            OcagentAddress:         "otel-collector.default:55678",
+            OcagentExporterEnabled: true,
+        }),
+    },
+},
+...
+```
+
+For Zipkin, an Envoy cluster corresponding to the Zipkin collector must exist. There are a few different options for creating this cluster:
```

Comment:
Isn't this exactly what ExtensionService is for? I'd prefer to see us reuse that for the second and third options.

Reply:
There are a couple challenges with using ExtensionService here:
- Zipkin doesn't necessarily use gRPC (though it can). Right now, ExtensionService pretty much assumes the service is gRPC.
- for OpenCensus, the API *is* gRPC, but Envoy's OC tracer does not allow using the Envoy gRPC client, only the Google gRPC client. As such, it can't be directly targeted at an Envoy cluster; it needs a gRPC address. I'm not sure off the top of my head how we'd make this work.

If we're committed to using ExtensionService here, then I can look more into what changes would be needed to make this work. TBH, I feel like this would be easier if instead of having a single generic ExtensionService type, we had `AuthorizationService`, `RateLimitService`, `TracingService`, etc., that maybe shared some common configuration but could also vary independently to support the variations among the different services. I know that's a bigger redesign though.

Reply:
> I feel like this would be easier if instead of having a single generic ExtensionService type, we had AuthorizationService, RateLimitService, TracingService, etc

+1

Reply:
I think the last time we discussed this, @skriss was going to work on an updated ExtensionService design (which may include separate *Service objects for each use case). Is that right?

Reply:
Yep, I've been thinking about it and starting to write up a doc.

Reply:
I think if we moved to a CRD per service type (ExtAuthService, RateLimitService, TracingService), we could make this work since each type could be significantly different - and tracing would need to be significantly different than the others given that it's not an xDS gRPC service, and, depending on the provider, does not make use of an Envoy cluster at all (which is effectively what an ExtensionService is right now). However, I think it's going to be difficult to move from the generic ExtensionService to a CRD per service type since it'll require what I believe is a breaking API change for ExtAuth (HTTPProxy currently has an implicitly-typed reference to an ExtensionService), and disruptive/breaking config file changes for rate limiting. If we wanted to do that, it might need to wait for a v2 API/v2 Contour.

Another option would be to add type-specific subfields to ExtensionService; however, because tracing is so different from the existing ExtensionService types, many of the existing fields in ExtensionService don't apply to tracing, so we end up with a messy API where many fields just don't apply in certain circumstances. I'm not a huge fan of this option given where we're starting from -- even though ExtensionService is "alpha", we don't want to break users if we can avoid it, which makes it harder to make in-place changes.

We could introduce just a TracingService CRD in the `v1alpha1` group which would give us a chance to experiment with having a CRD per service type, and then look to migrate the other existing service types to specific CRDs down the road. I like this approach in theory, though there are some challenges specifically around OpenCensus since it can't be configured more than once per Envoy lifetime (that would apply even if we used the config file).

Or, we could just use the config file as originally proposed.

Separately, since a main driver for defining an ExtensionService CRD to begin with was that it gave a way to report status, I think we should do that work for the existing service types. We could start simple by reporting whether or not a given ExtensionService has healthy Kubernetes Endpoints for its services, and then decide if we wanted to report more information from there.

Before going much further down any one of these paths, I'd appreciate some input from others.

Reply:
Thanks for that summary @skriss. I kind of expected that if we were going to look at separate objects, we would start with TracingService and then come back for the others later, I think that is fine. That's why these are alpha, after all. If the way tracing needs to work is that the config file references the TracingService rather than configuring it at an object level, I thin that's fine. Having the TracingService then makes it easier to operate tracing sinks, since you can flip to a new one, check everything is working, then flip back if necessary (although obviously it's not that easy for OpenCensus if it's per-Envoy-lifetime).

I don't like the idea of adding extra fields to ExtensionService because of what you said - although the structs may live on in the new CRDs, I think the use cases are distinct enough to need separate things. We just need to keep an eye on how many we end up with.

I agree that we need to do something with Status for ExtensionService, and that "Are there a nonzero number of Endpoints available" is a good start on a Ready conditrion.

For moving the existing things like ExternalAuth references, we could do something like:
- add GVK fields to the references, default to ExtensionService in one release.
- Later, change default to reference AuthService instead. This is a breaking change, but an easy fix for users, just specify the kind.
- Later, officially announce ExtensionService as deprecated and ask people to move to AuthorizationService.
Total timeframe for this, I would guess nine to twelve months, however many releases that is.